asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go

```diff
@@ -0,0 +1,584 @@
+	Owner          string `json:"owner"`
+	AppraisedValue int    `json:"appraisedValue"`
+}
+
+// QueryResult structure used for handling result of query
+type QueryResult struct {
+	Record              *asset
+	TxId                string    `json:"txId"`
+	Timestamp           time.Time `json:"timestamp"`
+	FetchedRecordsCount int       `json:"fetchedRecordsCount"`
+	Bookmark            string    `json:"bookmark"`
+}
+
+// ============================================================
+// CreateAsset - create a new asset, store into chaincode state
+// ============================================================
+func (t *SimpleChaincode) CreateAsset(ctx contractapi.TransactionContextInterface, ID, color, owner string, size, appraisedValue int) error {
+
+	// ==== Check if asset already exists ====
+	exists, err := t.AssetExists(ctx, ID)
+	if err != nil {
+		return fmt.Errorf("Failed to get asset: " + err.Error())
+	} else if exists {
+		return fmt.Errorf("This asset already exists: " + ID)
+	}
+
+	// ==== Create asset object and marshal to JSON ====
+	objectType := "asset"
+	asset := &asset{objectType, ID, color, size, owner, appraisedValue}
+	assetJSONasBytes, err := json.Marshal(asset)
+	if err != nil {
+		return err
+	}
+	//Alternatively, build the asset json string manually if you don't want to use struct marshalling
+	//assetJSONasString := `{"docType":"asset",  "name": "` + ID + `", "color": "` + color + `", "size": ` + strconv.Itoa(size) + `, "owner": "` + owner + `"}`
+	//assetJSONasBytes := []byte(str)
+
+	// === Save asset to state ===
+	err = ctx.GetStub().PutState(ID, assetJSONasBytes)
+	if err != nil {
+		return err
+	}
+
+	//  ==== Index the asset to enable color-based range queries, e.g. return all blue assets ====
+	//  An 'index' is a normal key/value entry in state.
+	//  The key is a composite key, with the elements that you want to range query on listed first.
+	//  In our case, the composite key is based on indexName~color~name.
+	//  This will enable very efficient state range queries based on composite keys matching indexName~color~*
+	indexName := "color~name"
+	colorNameIndexKey, err := ctx.GetStub().CreateCompositeKey(indexName, []string{asset.Color, asset.ID})
+	if err != nil {
+		return err
+	}
+	//  Save index entry to state. Only the key name is needed, no need to store a duplicate copy of the asset.
+	//  Note - passing a 'nil' value will effectively delete the key from state, therefore we pass null character as value
+	value := []byte{0x00}
+	ctx.GetStub().PutState(colorNameIndexKey, value)
+
+	// ==== Asset saved and indexed. Return success ====
+	return nil
+}
+
+// ===============================================
+// ReadAsset - read a asset from chaincode state
+// ===============================================
+func (t *SimpleChaincode) ReadAsset(ctx contractapi.TransactionContextInterface, name string) (*asset, error) {
+	assetJSON, err := ctx.GetStub().GetState(name) //get the asset from chaincode state
```

Comment:
You can remove many of these obvious comments, comments are only needed for non-obvious things.